### PR TITLE
Batched pairfunction cleanup

### PR DESCRIPTION
**Pull request by CompRhys**
Replace all the old classical pair potential backends with general backend, make this differentiable to enable `diff_sim.py` to work.

### CODE REVIEW

**Comment by CompRhys** [comment]
The PR was looking so clean until these tests broke. They all use LJ models and have numerical consistency issues. Not sure what the bug might be :(

File: examples/tutorials/diff_sim.py [hunk 1]
```diff
@@ -9,17 +9,19 @@
 # </details>
 
 # %%
+import typing
```
Comment (abhijeetgangan) [inline examples/tutorials/diff_sim.py:12]: You can remove the example because it seems out of scope for torchsim.

Comment (orionarcher) [reply examples/tutorials/diff_sim.py:12]: I think having a differentiable example for classical potentials is nice

File: tests/test_integrators.py [hunk 2]
```diff
@@ -339,7 +339,7 @@ def test_nvt_nose_hoover(ar_double_sim_state: ts.SimState, lj_model: LennardJone
     temperatures_list = [t.tolist() for t in temperatures_tensor.T]
     assert torch.allclose(
         temperatures_tensor[-1],
-        torch.tensor([299.9910, 299.6800], dtype=dtype),
+        torch.tensor([300.0096, 299.7024], dtype=dtype),
```
Comment (orionarcher) [inline tests/test_integrators.py:342]: Is this degree of precision justified?

File: torch_sim/models/pair_potential.py [hunk 3]
```diff
@@ -9,26 +9,24 @@
 (e.g. the asymmetric particle-life interaction) that cannot be expressed as the
 gradient of a scalar energy.
```
Comment (orionarcher) [inline torch_sim/models/pair_potential.py:10]: More documentation of new design perhaps?

File: tests/test_integrators.py (hunk 2 above)
Comment (CompRhys) [reply tests/test_integrators.py:342]: This was a magic seeded test and the change in the backend caused the numbers to jump. We could say +/- 1 on the all close and remove the seed but then we would have to contend with it potentially becoming flaky
